proxyserver/middleware/copy.go

```diff
@@ -0,0 +1,433 @@
+}
+
+func (c *copyMiddleware) getSourceObject(object string, request *http.Request) (io.ReadCloser, http.Header, int) {
+	ctx := GetProxyContext(request)
+	subRequest, err := ctx.NewSubRequest(request, "GET", object, nil)
+	if err != nil {
+		ctx.Logger.Error("getSourceObject GET error", zap.Error(err))
+		return nil, nil, 400
+	}
+
+	pipeReader, pipeWriter := io.Pipe()
+	done := make(chan bool)
+	writer := NewPipeResponseWriter(pipeWriter, done, ctx)
+	go func() {
+		c.next.ServeHTTP(writer, subRequest)
+		writer.Close()
+	}()
+	<-done
+
+	return pipeReader, writer.Header(), writer.status
+}
+
+func (c *copyMiddleware) handlePostAsCopy(writer *CopyWriter, request *http.Request) {
+	request.Method = "PUT"
+	writer.postAsCopy = true
+
+	request.URL.Path = fmt.Sprintf("/v1/%s/%s/%s", writer.accountName, writer.containerName, writer.objectName)
+	request.ContentLength = 0
+	request.Header.Del("Range")
+	objectPath := fmt.Sprintf("/%s/%s", writer.containerName, writer.objectName)
+	request.Header.Set("X-Copy-From", common.Urlencode(objectPath))
+
+	values, err := url.ParseQuery(request.URL.RawQuery)
+	if err != nil {
+		srv.StandardResponse(writer, 400)
+	}
+	values.Set("Multipart-Manifest", "get")
+	request.URL.RawQuery = values.Encode()
+
+	c.handlePut(writer, request)
+}
+
+func (c *copyMiddleware) handleCopy(writer *CopyWriter, request *http.Request) {
+	if request.Header.Get("Destination") == "" {
+		// FIXME.
+		// swift has: body='Destination header required'
+		srv.StandardResponse(writer, 412)
+	}
+	destAccount := writer.getDestAccountName(request)
+	destContainer, destObject, err := getDestContainerObjectName(request)
+	if err != nil {
+		srv.StandardResponse(writer, 412)
+		return
+	}
+	source := fmt.Sprintf("%s/%s", writer.containerName, writer.objectName)
+	writer.containerName = destContainer
+	writer.objectName = destObject
+
+	request.Method = "PUT"
+	request.URL.Path = fmt.Sprintf("/v1/%s/%s/%s", destAccount, destContainer, destObject)
+	request.ContentLength = 0
+	request.Header.Set("X-Copy-From", common.Urlencode(source))
+	request.Header.Del("Destination")
+
+	c.handlePut(writer, request)
+}
+
+func RemoveItemsWithPrefix(header http.Header, prefix string) {
+	for k := range header {
+		if strings.HasPrefix(k, prefix) {
+			header.Del(k)
+		}
+	}
+}
+
+func CopyItemsWithPrefix(dest, src http.Header, prefix string) {
+	for k, v := range src {
+		if strings.HasPrefix(k, prefix) {
+			dest.Del(k)
+			for _, v1 := range v {
+				dest.Add(k, v1)
+			}
+		}
+	}
+}
+
+func CopyItems(dest, src http.Header) {
+	for k, v := range src {
+		dest.Del(k)
+		for _, v1 := range v {
+			dest.Add(k, v1)
+		}
+	}
+}
+
+func CopyMetaItems(dest, src http.Header) {
+	CopyItemsWithPrefix(dest, src, "X-Object-Meta-")
+	CopyItemsWithPrefix(dest, src, "X-Object-Sysmeta-")
+	CopyItemsWithPrefix(dest, src, "X-Object-Transient-Sysmeta-")
+	dest.Set("X-Delete-At", src.Get("X-Delete-At"))
+}
+
+func CopyItemsExclude(dest, src http.Header, exclude []string) {
```

Comment:
i just meant you could just test for membership in the map instead of walking through the array